sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/V1Beta3Helper.java

```diff
@@ -0,0 +1,486 @@
+  static final long DEFAULT_BUNDLE_SIZE_BYTES = 64 * 1024 * 1024;
+
+  // A lower bound on the number of splits for a query
+  static final int NUM_QUERY_SPLITS_MIN = 12;
+
+  /**
+   * Builds a datastore client for the given {@param pipelineOptions} and {@param projectId}.
+   */
+  static Datastore getDatastore(DatastoreFactory datastoreFactory, PipelineOptions pipelineOptions,
+      String projectId) {
+    DatastoreOptions.Builder builder =
+        new DatastoreOptions.Builder()
+            .projectId(projectId)
+            .initializer(
+                new RetryHttpRequestInitializer()
+            );
+
+    Credential credential = pipelineOptions.as(GcpOptions.class).getGcpCredential();
+    if (credential != null) {
+      builder.credential(credential);
+    }
+    return datastoreFactory.create(builder.build());
+  }
+
+  /**
+   * Computes the number of splits to be performed on the given {@param query} by
+   * querying the estimated size from datastore. We use {@code DEFAULT_BUNDLE_SIZE_BYTES} as the
+   * bundle size.
+   */
+  static int getEstimatedNumSplits(Datastore datastore, Query query, String namespace) {
+    int numSplits;
+    try {
+      long estimatedSizeBytes = getEstimatedSizeBytes(datastore, query, namespace);
+      numSplits = (int) Math.min(Integer.MAX_VALUE,
+          Math.round(((double) estimatedSizeBytes) / DEFAULT_BUNDLE_SIZE_BYTES));
+    } catch (Exception e) {
+      LOG.warn("Failed the fetch estimatedSizeBytes for query: {}", query, e);
+      // Fallback in case estimated size is unavailable.
+      numSplits = NUM_QUERY_SPLITS_MIN;
+    }
+    return Math.max(numSplits, NUM_QUERY_SPLITS_MIN);
+  }
+
+  /**
+   * Get the estimated size of the data returned by the given query.
+   *
+   * Datastore provides no way to get a good estimate of how large the result of a query
+   * entity kind being queried, using the __Stat_Kind__ system table, assuming exactly 1 kind
+   * is specified in the query.
+   *
+   * See https://cloud.google.com/datastore/docs/concepts/stats
+   */
+  static long getEstimatedSizeBytes(Datastore datastore, Query query, String namespace)
+      throws DatastoreException {
+    String ourKind = query.getKind(0).getName();
+    Query.Builder queryBuilder = Query.newBuilder();
+    if (namespace == null) {
+      queryBuilder.addKindBuilder().setName("__Stat_Kind__");
+    } else {
+      queryBuilder.addKindBuilder().setName("__Ns_Stat_Kind__");
+    }
+    queryBuilder.setFilter(makeFilter("kind_name", EQUAL, makeValue(ourKind).build()));
+
+    // Get the latest statistics
+    queryBuilder.addOrder(makeOrder("timestamp", DESCENDING));
+    queryBuilder.setLimit(Int32Value.newBuilder().setValue(1));
+
+    RunQueryRequest request = makeRequest(queryBuilder.build(), namespace);
+
+    long now = System.currentTimeMillis();
+    RunQueryResponse response = datastore.runQuery(request);
+    LOG.info("Query for per-kind statistics took {}ms", System.currentTimeMillis() - now);
```

Comment:
i know this was there, but I'd drop or at least lower to debug level. This is probably legacy code when someone was concerned about RPC latency.